bin/prover-client/src/operators/mod.rs

```diff
@@ -0,0 +1,152 @@
+        &self,
+        params: Self::Params,
+        task_tracker: Arc<Mutex<TaskTracker>>,
+        db: &ProofDb,
+    ) -> Result<Vec<ProofKey>, ProvingTaskError>;
+
+    /// Fetches the input required for the proof computation.
+    ///
+    /// # Arguments
+    /// - `task_id`: The key representing the proof task.
+    /// - `db`: A reference to the proof database.
+    ///
+    /// # Returns
+    /// The input required by the prover for the specified task.
+    async fn fetch_input(
+        &self,
+        task_id: &ProofKey,
+        db: &ProofDb,
+    ) -> Result<<Self::Prover as ZkVmProver>::Input, ProvingTaskError>;
+
+    /// Executes the proof computation for the specified task.
+    ///
+    /// # Arguments
+    /// - `task_id`: The key representing the proof task.
+    /// - `db`: A reference to the proof database.
+    ///
+    /// # Returns
+    /// An empty result if the proof computation is successful.
+    #[instrument(skip(self, db), fields(task_id = ?task_id))]
+    async fn prove(&self, task_id: &ProofKey, db: &ProofDb) -> Result<(), ProvingTaskError> {
+        info!("Starting proof generation");
+
+        let input = self
+            .fetch_input(task_id, db)
+            .await
+            .inspect_err(|e| error!(?e, "Failed to fetch input"))?;
+
+        let proof_res = match task_id.host() {
+            ProofZkVm::Native => {
+                let host = hosts::native::get_host(task_id.context());
+                <Self::Prover as ZkVmProver>::prove(&input, &host)
+            }
+            ProofZkVm::SP1 => {
+                #[cfg(feature = "sp1")]
+                {
+                    let host = hosts::sp1::get_host(task_id.context());
+                    <Self::Prover as ZkVmProver>::prove(&input, host)
+                }
+                #[cfg(not(feature = "sp1"))]
+                {
+                    panic!("The `sp1` feature is not enabled. Enable the feature to use SP1 functionality.");
+                }
+            }
+            ProofZkVm::Risc0 => {
+                #[cfg(feature = "risc0")]
+                {
+                    let host = hosts::risc0::get_host(task_id.context());
+                    <Self::Prover as ZkVmProver>::prove(&input, host)
+                }
+                #[cfg(not(feature = "risc0"))]
+                {
+                    panic!("The `risc0` feature is not enabled. Enable the feature to use Risc0 functionality.");
+                }
+            }
+            _ => {
+                panic!("Unsupported ZkVm");
+            }
+        };
+
+        match &proof_res {
+            Ok(_) => {
+                info!("Proof generated successfully")
+            }
+            Err(e) => {
+                error!(?e, "Failed to generate proof")
+            }
+        }
+
+        let proof = proof_res.map_err(ProvingTaskError::ZkVmError)?;
+
+        db.put_proof(*task_id, proof)
+            .map_err(ProvingTaskError::DatabaseError)?;
+
+        Ok(())
```

Comment:
I know we discussed this. I gave some thought. I still don't like implementation details to be hardcoded in abstractions like traits.

This should be moved into the impls themselves:

```suggestion
    /// Executes the proof computation for the specified task.
    ///
    /// # Arguments
    /// - `task_id`: The key representing the proof task.
    /// - `db`: A reference to the proof database.
    ///
    /// # Returns
    /// An empty result if the proof computation is successful.
    #[instrument(skip(self, db), fields(task_id = ?task_id))]
    async fn prove(&self, task_id: &ProofKey, db: &ProofDb) -> Result<(), ProvingTaskError>;
```

Reply:
I agree with this.  This is not what default impls are designed for.  Maybe we should add a lint for that to annoy you and make you rethink it.

Reply:
Have tried to solve this. 
PTAL @delbonis @storopoli